File: Clients/Node/Admin/src/signalr/AdminNotificationHubClient.ts
import { HubConnection } from '@microsoft/signalr';
import { BaseSignalRConnection } from './BaseSignalRConnection';
import { 
  SignalREndpoints, 
  VirtualKeyEvent,
  ConfigurationChangeEvent,
  AdminNotificationEvent,
  IAdminNotificationHubClient
} from '../models/signalr';

/**
 * SignalR client for admin notifications
 */
export class AdminNotificationHubClient extends BaseSignalRConnection implements IAdminNotificationHubClient {
  protected get hubPath(): string {
    return SignalREndpoints.AdminNotifications;
  }

  private virtualKeyCallbacks: ((event: VirtualKeyEvent) => void)[] = [];
  private configChangeCallbacks: ((event: ConfigurationChangeEvent) => void)[] = [];
  private adminNotificationCallbacks: ((event: AdminNotificationEvent) => void)[] = [];
  private initialProviderHealthCallbacks: ((data: unknown) => void)[] = [];

  /**
   * Configures event handlers for the admin notification hub
   */
  protected configureHubHandlers(connection: HubConnection): void {
    // Virtual key event handler
    connection.on('VirtualKeyEvent', (event: VirtualKeyEvent) => {
      this.virtualKeyCallbacks.forEach(callback => {
        try {
          callback(event);
        } catch (error) {
          console.error('Error in virtual key event callback:', error);
        }
      });
    });

    // Configuration change handler
    connection.on('ConfigurationChanged', (event: ConfigurationChangeEvent) => {
      this.configChangeCallbacks.forEach(callback => {
        try {
          callback(event);
        } catch (error) {
          console.error('Error in configuration change callback:', error);
        }
      });
    });

    // Admin notification handler
    connection.on('AdminNotification', (event: AdminNotificationEvent) => {
      this.adminNotificationCallbacks.forEach(callback => {
        try {
          callback(event);
        } catch (error) {
          console.error('Error in admin notification callback:', error);
        }
      });
    });

    // Initial provider health handler (sent on connection)
    connection.on('InitialProviderHealth', (data: unknown) => {
      this.initialProviderHealthCallbacks.forEach(callback => {
        try {
          callback(data);
        } catch (error) {
          console.error('Error in initial provider health callback:', error);
        }
      });
    });

    // Also handle other server events that might be sent
    connection.on('Error', (data: { message: string }) => {
      console.error('Server error:', data.message);
    });

    connection.on('ProviderHealthStatus', (data: unknown) => {
      console.log('Provider health status received:', data);
    });

    connection.on('ProviderHealthRefreshed', (data: unknown) => {
      console.log('Provider health refreshed:', data);
    });
  }

  /**
   * Subscribe to virtual key events
   */
  onVirtualKeyEvent(callback: (event: VirtualKeyEvent) => void): void {
    this.virtualKeyCallbacks.push(callback);
  }

  /**
   * Subscribe to configuration changes
   */
  onConfigurationChange(callback: (event: ConfigurationChangeEvent) => void): void {
    this.configChangeCallbacks.push(callback);
  }

  /**
   * Subscribe to admin notifications
   */
  onAdminNotification(callback: (event: AdminNotificationEvent) => void): void {
    this.adminNotificationCallbacks.push(callback);
  }

  /**
   * Subscribe to initial provider health updates
   */
  onInitialProviderHealth(callback: (data: unknown) => void): void {
    this.initialProviderHealthCallbacks.push(callback);
  }

  /**
   * Subscribe to notifications for a specific virtual key
   */
  async subscribeToVirtualKey(virtualKeyId: number): Promise<void> {
    await this.invoke('SubscribeToVirtualKey', virtualKeyId);
  }

  /**
   * Unsubscribe from notifications for a specific virtual key
   */
  async unsubscribeFromVirtualKey(virtualKeyId: number): Promise<void> {
    await this.invoke('UnsubscribeFromVirtualKey', virtualKeyId);
  }

  /**
   * Subscribe to notifications for a specific provider
   */
  async subscribeToProvider(providerName: string): Promise<void> {
    await this.invoke('SubscribeToProvider', providerName);
  }

  /**
   * Unsubscribe from notifications for a specific provider
   */
  async unsubscribeFromProvider(providerName: string): Promise<void> {
    await this.invoke('UnsubscribeFromProvider', providerName);
  }

  /**
   * Request a refresh of provider health status
   */
  async refreshProviderHealth(): Promise<void> {
    await this.invoke('RefreshProviderHealth');
  }

  /**
   * Acknowledge a notification as read
   */
  async acknowledgeNotification(notificationId: string): Promise<void> {
    await this.invoke('AcknowledgeNotification', notificationId);
  }

  /**
   * Clear all callbacks
   */
  clearCallbacks(): void {
    this.virtualKeyCallbacks = [];
    this.configChangeCallbacks = [];
    this.adminNotificationCallbacks = [];
    this.initialProviderHealthCallbacks = [];
  }

  /**
   * Remove a specific virtual key callback
   */
  removeVirtualKeyCallback(callback: (event: VirtualKeyEvent) => void): void {
    const index = this.virtualKeyCallbacks.indexOf(callback);
    if (index > -1) {
      this.virtualKeyCallbacks.splice(index, 1);
    }
  }

  /**
   * Remove a specific configuration change callback
   */
  removeConfigChangeCallback(callback: (event: ConfigurationChangeEvent) => void): void {
    const index = this.configChangeCallbacks.indexOf(callback);
    if (index > -1) {
      this.configChangeCallbacks.splice(index, 1);
    }
  }

  /**
   * Remove a specific admin notification callback
   */
  removeAdminNotificationCallback(callback: (event: AdminNotificationEvent) => void): void {
    const index = this.adminNotificationCallbacks.indexOf(callback);
    if (index > -1) {
      this.adminNotificationCallbacks.splice(index, 1);
    }
  }

  /**
   * Get the number of active callbacks
   */
  getActiveCallbackCount(): number {
    return this.virtualKeyCallbacks.length + 
           this.configChangeCallbacks.length + 
           this.adminNotificationCallbacks.length +
           this.initialProviderHealthCallbacks.length;
  }

  /**
   * Dispose the client
   */
  async dispose(): Promise<void> {
    this.clearCallbacks();
    await super.dispose();
  }
}

File: Clients/DotNet/ConduitLLM.AdminClient/src/SignalR/AdminNotificationHubClient.cs
using Microsoft.AspNetCore.SignalR.Client;
using Microsoft.Extensions.Logging;
using ConduitLLM.AdminClient.Models;

namespace ConduitLLM.AdminClient.SignalR;

/// <summary>
/// SignalR hub client for admin notifications including virtual key events,
/// configuration changes, and system notifications.
/// </summary>
public class AdminNotificationHubClient : BaseSignalRConnection
{
    /// <summary>
    /// Event raised when a virtual key event occurs.
    /// </summary>
    public event Action<VirtualKeyEvent>? OnVirtualKeyEvent;
    
    /// <summary>
    /// Event raised when a configuration change occurs.
    /// </summary>
    public event Action<ConfigurationChangeEvent>? OnConfigurationChange;
    
    /// <summary>
    /// Event raised when an admin notification is received.
    /// </summary>
    public event Action<AdminNotificationEvent>? OnAdminNotification;

    /// <summary>
    /// Initializes a new instance of the AdminNotificationHubClient class.
    /// </summary>
    /// <param name="baseUrl">Base URL of the Conduit Admin API.</param>
    /// <param name="masterKey">Master key for authentication.</param>
    /// <param name="logger">Optional logger instance.</param>
    public AdminNotificationHubClient(string baseUrl, string masterKey, ILogger<AdminNotificationHubClient>? logger = null)
        : base(baseUrl, masterKey, logger)
    {
    }

    /// <summary>
    /// Gets the hub path for admin notifications.
    /// </summary>
    protected override string HubPath => SignalREndpoints.AdminNotificationsHub;

    /// <summary>
    /// Configures hub-specific event handlers.
    /// </summary>
    /// <param name="connection">The hub connection to configure.</param>
    protected override void ConfigureHubHandlers(HubConnection connection)
    {
        // Virtual key events
        connection.On<VirtualKeyEvent>("VirtualKeyEvent", (eventData) =>
        {
            _logger?.LogDebug("Received virtual key event: {EventType} for key {VirtualKeyId}", 
                eventData.EventType, eventData.VirtualKeyId);
            OnVirtualKeyEvent?.Invoke(eventData);
        });

        // Configuration changes
        connection.On<ConfigurationChangeEvent>("ConfigurationChange", (eventData) =>
        {
            _logger?.LogDebug("Received configuration change: {Category}.{Setting}", 
                eventData.Category, eventData.Setting);
            OnConfigurationChange?.Invoke(eventData);
        });

        // Admin notifications
        connection.On<AdminNotificationEvent>("AdminNotification", (notification) =>
        {
            _logger?.LogInformation("Received admin notification: [{Type}] {Title}", 
                notification.Type, notification.Title);
            OnAdminNotification?.Invoke(notification);
        });

        _logger?.LogDebug("Configured handlers for AdminNotificationHub");
    }

    /// <summary>
    /// Subscribes to events for a specific virtual key.
    /// </summary>
    /// <param name="virtualKeyId">The virtual key ID to subscribe to.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task SubscribeToVirtualKey(int virtualKeyId, CancellationToken cancellationToken = default)
    {
        await InvokeAsync("SubscribeToVirtualKey", new object[] { virtualKeyId }, cancellationToken);
        _logger?.LogInformation("Subscribed to virtual key events for key ID {VirtualKeyId}", virtualKeyId);
    }

    /// <summary>
    /// Unsubscribes from events for a specific virtual key.
    /// </summary>
    /// <param name="virtualKeyId">The virtual key ID to unsubscribe from.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task UnsubscribeFromVirtualKey(int virtualKeyId, CancellationToken cancellationToken = default)
    {
        await InvokeAsync("UnsubscribeFromVirtualKey", new object[] { virtualKeyId }, cancellationToken);
        _logger?.LogInformation("Unsubscribed from virtual key events for key ID {VirtualKeyId}", virtualKeyId);
    }

    /// <summary>
    /// Subscribes to events for a specific provider.
    /// </summary>
    /// <param name="providerName">The provider name to subscribe to.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task SubscribeToProvider(string providerName, CancellationToken cancellationToken = default)
    {
        await InvokeAsync("SubscribeToProvider", new object[] { providerName }, cancellationToken);
        _logger?.LogInformation("Subscribed to provider events for {ProviderName}", providerName);
    }

    /// <summary>
    /// Unsubscribes from events for a specific provider.
    /// </summary>
    /// <param name="providerName">The provider name to unsubscribe from.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task UnsubscribeFromProvider(string providerName, CancellationToken cancellationToken = default)
    {
        await InvokeAsync("UnsubscribeFromProvider", new object[] { providerName }, cancellationToken);
        _logger?.LogInformation("Unsubscribed from provider events for {ProviderName}", providerName);
    }

    /// <summary>
    /// Requests a refresh of provider health status.
    /// </summary>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task RefreshProviderHealth(CancellationToken cancellationToken = default)
    {
        await InvokeAsync("RefreshProviderHealth", Array.Empty<object>(), cancellationToken);
        _logger?.LogInformation("Requested provider health refresh");
    }

    /// <summary>
    /// Acknowledges receipt of a notification.
    /// </summary>
    /// <param name="notificationId">The notification ID to acknowledge.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    public async Task AcknowledgeNotification(string notificationId, CancellationToken cancellationToken = default)
    {
        await InvokeAsync("AcknowledgeNotification", new object[] { notificationId }, cancellationToken);
        _logger?.LogDebug("Acknowledged notification {NotificationId}", notificationId);
    }
}